src/stirling/source_connectors/socket_tracer/bcc_bpf/grpc_c_trace.c

```diff
@@ -0,0 +1,1557 @@
+                &current_linked_mdelem,
+                sizeof(current_linked_mdelem),
+                (void *)(current_linked_mdelem + 0x8)))
+        {
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+/*
+ * @brief   Handle metadata being fully received.
+ * @detailed
+ *          This is the immediate handler to the
+ *          "grpc_chttp2_maybe_complete_recv_initial_metadata" and
+ *          "grpc_chttp2_maybe_complete_recv_trailing_metadata" functions.
+ *          This fires a perf buffer event for data being read, without actual data
+ *          (only headers).
+ *
+ * @param   ctx             The context of the probe.
+ * @param   is_initial      Whether handling initial metadata being received or trailing
+ *                          metadata being received.
+ *
+ * @return  0 on success.
+ *          Otherwise on failure.
+ */
+static inline int handle_maybe_complete_recv_metadata(struct pt_regs * ctx, const bool is_initial)
+{
+    struct grpc_c_event_data_t * read_data = initiate_empty_grpc_event_data();
+    if (NULL == read_data)
+    {
+        return -1;
+    }
+
+    read_data->direction = GRPC_C_EVENT_DIRECTION_INCOMING;
+    grpc_metadata_batch * metadata_batch = NULL;
+    struct grpc_c_metadata_t * metadata = NULL;
+    u32 key = 0;
+    u32 initial_metadata_buffer_offset = 0;
+    u32 trailing_metadata_buffer_offset = 0;
+    u32 offset = 0;
+    grpc_chttp2_stream * stream_ptr = NULL;
+    grpc_chttp2_transport * transport_ptr = NULL;
+
+    u32 pid = bpf_get_current_pid_tgid() >> 32;
+    u32 fd = 0;
+    u64 version = lookup_version(pid);
+    if (GRPC_C_VERSION_UNSUPPORTED == version)
+    {
+        return 0;
+    }
+    read_data->timestamp = bpf_ktime_get_ns();
+    transport_ptr = (grpc_chttp2_transport *)PT_REGS_PARM1(ctx);
+    stream_ptr = (grpc_chttp2_stream *)PT_REGS_PARM2(ctx);
+    read_data->stack_id = socket_stack_traces.get_stackid(
+        ctx, BPF_F_USER_STACK | BPF_F_REUSE_STACKID);
+
+    if (NULL == transport_ptr || NULL == stream_ptr)
+    {
+        return -1;
+    }
+
+    if (0 != get_stream_id((grpc_chttp2_stream *)stream_ptr, &read_data->stream_id, version))
+    {
+        return -1;
+    }
+
+    if (0 != get_fd_from_transport((grpc_chttp2_transport *)transport_ptr, &fd))
+    {
+        return -1;
+    }
+
+    struct conn_info_t* conn_info = get_or_create_conn_info(pid, fd);
+    if (NULL == conn_info)
+    {
+        return -1;
+    }
+    read_data->conn_id = conn_info->conn_id;
+
+    switch (version)
+    {
+        case GRPC_C_V1_19_0:
+            initial_metadata_buffer_offset = 0x1e0;
+            trailing_metadata_buffer_offset = 0x2d8;
+            break;
+        case GRPC_C_V1_24_1:
+            // Not supported yet.
+            break;
+        case GRPC_C_V1_33_2:
+            initial_metadata_buffer_offset = 0x330;
+            trailing_metadata_buffer_offset = 0x570;
+            break;
+        case GRPC_C_V1_41_1:
+            // Not supported yet.
```

Comment:
About versions - we should probably make some product effort to understand what versions are most important and focus our efforts there. These 2 can be very easily added, and I suppose everything between 1.19 to 1.41 as well.

Reply:
about return value - In the entire code I treated "unsupported version" with return value 0 and not -1 (You can see it after every call to lookup_version).
This case should probably match the other ones.
So, If you see "unsupported" as an error (-1 instead of 0), I'll change it everywhere. Let me know what you decide.